File: account-server-golang/server_api/READ_GAME_SERVERS.go
package server_api

import (
	"account-server/database"
	"account-server/gameservers"
	"account-server/utils"
	"encoding/json"
	"errors"
	"fmt"
	"time"
)

type request_READ_GAME_SERVERS struct {
	Request_id string
	Session_id string
}

func READ_GAME_SERVERS(message []byte, sendResponse func([]byte)) error {

	var r request_READ_GAME_SERVERS

	// Comprueba que estan todos los parametros requeridos.
	{
		var m map[string]any
		err := json.Unmarshal(message, &m)
		if err != nil {
			return err
		}
		if !utils.ContainsAllKeys(m, []string{"request_id", "session_id"}) {
			err = errors.New("READ_GAME_SERVERS bad input parameters")
			return err
		}

		if _, ok := m["require_test_instance"]; ok { // Evita que las requests, que indican require_test_instance, se ejecuten en una instancia del account server que no es de test.
			if !database.IsTestInstance() {
				return errors.New("require_test_instance")
			}
		}
	}

	err := json.Unmarshal(message, &r)
	if err != nil {
		return err
	}

	var account dict

	// Comprueba si existe una cuenta con esa session_id
	{
		target := []string{"session_id", r.Session_id}
		responseDict, err := database.ReadAccount(target, "session_expire guest_account_id")
		if err != nil {
			return err
		}
		account = responseDict
	}

	responseDict := dict{}

	if account != nil {

		sessionExpired := false

		if account["guest_account_id"] != nil {
			// Para las cuentas guest no se aplica session_expire
		} else {
			// Comprueba que la session no esta caducada
			expireTimeStr := account["session_expire"].(string) // Sera "" si session_expire no esta en account
			if expireTimeStr == "" {
				panic(0)
			}
			expireTime, err := time.Parse(time.RFC3339, expireTimeStr)
			if err != nil {
				panic(0)
			}
			sessionExpired = expireTime.Before(time.Now())
		}

		if !sessionExpired {

			// Read game servers
			{
				gservers := gameservers.GetGameServers()

				gameServersInfoList := []dict{}
				for _, gsie := range gservers {
					gameServersInfoList = append(gameServersInfoList, dict{
						"lobby_id":             gsie.Lobby_id,
						"server_name":          gsie.Server_name,
						"ipv4_address":         gsie.Ipv4_address + ":" + fmt.Sprint(gsie.Port),
						"address":              gsie.Address + ":" + fmt.Sprint(gsie.Port),
						"region":               gsie.Region,
						"player_count":         gsie.PlayerCount,
						"max_player_count":     gsie.Max_rooms * gsie.Max_players_per_room,
						"max_rooms":            gsie.Max_rooms,
						"max_players_per_room": gsie.Max_players_per_room,
						"game":                 dict{"game_name": gsie.GameName, "map_name": gsie.MapName},
						// "per_room_player_count": gsie.RoomPlayerCount,
					})
				}

				responseDict["session_id"] = r.Session_id
				responseDict["game_servers"] = gameServersInfoList
			}

		} else {
			responseDict["error"] = "Session is expired."
		}

	} else {
		responseDict["error"] = "No account found with this session_id."
	}

	// Envia la respuesta

	responseDict["request_id"] = r.Request_id

	response, err := json.Marshal(responseDict)
	if err != nil {
		return err
	}
	sendResponse(response)

	return nil
}

File: account-server/server_api/READ_GAME_SERVERS.py
from datetime import datetime
from utils import str_to_date


# READ_GAME_SERVERS
# 1. El cliente envia READ_GAME_SERVERS(session_id) al servidor.
# 2. El servidor comprueba que existe esa session_id y no esta caducada.
# 3. El servidor responde con informacion sobre los game servers.
async def READ_GAME_SERVERS(req, ws, send_message, accounts_cache, gameservers_info):

    # Aqui se debe comprobar que la req es correcta y en caso contrario se ignora sin enviar nada al cliente.
    if not all(k in req for k in ('request_id', 'session_id')):
        print("ERROR: READ_GAME_SERVERS bad input parameters.")
        return

    error_response = {
        "error": "Invalid ticket.",
        "request_id": req['request_id']
    }  

    valid = False

    ### Comprueba si existe esa session_id

    session_id = req["session_id"]

    query = {"session_id": session_id}
    projection = 'session_expire guest_account_id'
    response = await accounts_cache.read_account(req, ws, send_message, query, projection)
    if response is None:
        return

    ### Comprueba si la session esta caducada

    if 'session_expire' in response:
        # session_expire_date = datetime_from_json_date_str(response['session_expire'])
        # if session_expire_date >= datetime.datetime.now():
        if 'guest_account_id' in response or str_to_date(response['session_expire']) >= datetime.now():
            del response['session_expire']
            response['session_id'] = session_id
            response["game_servers"] = [{
                'lobby_id': k,
                'server_name': x['server_name'],
                'address': x['ipv4_address'] + ":" + str(x['port']),
                'player_count': len(x['players']),
                'max_player_count': x['max_player_count'],
                'game': x['game']}
                for k, x in gameservers_info['instances'].items()]
            response["request_id"] = req['request_id']

            await send_message(response, ws)  # send response
            valid = True
        else:
            error_response["error"] = "Session is expired."
    else:
        error_response["error"] = "No account found."

    ### Envia error al cliente.

    if not valid:
        await send_message(error_response, ws)

    

    

